python/xorbits/core/execution.py

```diff
@@ -68,16 +67,38 @@ def need_to_execute(ref: DataRef) -> bool:
     )
 
 
-def _collect_user_ns_refs() -> dict[int, DataRef]:
+def _is_in_final_results(ref: DataRef, results: list[DataRef]):
+    mars_entity = _get_mars_entity(ref)
+    result_mars_entities = [_get_mars_entity(result) for result in results]
+    for result_mars_entity in result_mars_entities:
+        stack = [result_mars_entity]
+        while stack:
+            e = stack.pop()
+            if e.key == mars_entity.key:
+                return True
+            stack.extend(e.inputs)
```

Comment:
We can only add those inputs that are not executed before? In order to save time.

Reply:
Good point. Fixed.